File: src/main/java/com/jackmeng/stl/stl_Spline.java
package com.jackmeng.stl;

import java.util.List;

public class stl_Spline
{
    private final List<Double> x;
    private final List<Double> y;
    private final int n;
    private final double[] z;

    public stl_Spline(List<Double> x, List<Double> y)
    {
        this.x = x;
        this.y = y;
        this.n = x.size();
        double[] h = new double[n];
        double[] b = new double[n];
        double[] u = new double[n];
        double[] v = new double[n];
        this.z = new double[n + 1];
        for (int i = 0; i < n - 1; i++)
        {
            h[i] = x.get(i + 1) - x.get(i);
            b[i] = (y.get(i + 1) - y.get(i)) / h[i];
        }
        u[1] = 2 * (h[0] + h[1]);
        v[1] = 6 * (b[1] - b[0]);
        for (int i = 2; i < n - 1; i++)
        {
            u[i] = 2 * (h[i] + h[i - 1]) - h[i - 1] * h[i - 1] / u[i - 1];
            v[i] = 6 * (b[i] - b[i - 1]) - h[i - 1] * v[i - 1] / u[i - 1];
        }
        z[n] = 0;
        for (int i = n - 2; i > 0; i--)
            z[i] = (v[i] - h[i] * z[i + 1]) / u[i];
        z[0] = 0;
    }

    public double interpolate(double t)
    {
        int i = 0;
        int j = n - 1;
        while (j - i > 1) {
            int k = (i + j) / 2;
            if (x.get(k) > t)
                j = k;
            else
                i = k;
        }
        double dx = t - x.get(i);
        return y.get(i) + dx * (z[i] / 2 + dx * (z[j] - z[i]) / (6 * (x.get(j) - x.get(i))) + (y.get(j) - y.get(i)) / (x.get(j) - x.get(i)));
    }
}

File: src/main/java/com/jackmeng/stl/stl_Spline.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

class stl_Spline(private val x:List<Double> , private val y:List<Double>)
{
	private val n:Int=x.size
	private val z:DoubleArray
	
	init
	{
		val h=DoubleArray(n)
		val b=DoubleArray(n)
		val u=DoubleArray(n)
		val v=DoubleArray(n)
		z=DoubleArray(n+1)
		for (i in 0 until n-1)
		{
			h[i]=x[i+1]-x[i]
			b[i]=(y[i+1]-y[i])/h[i]
		}
		u[1]=2*(h[0]+h[1])
		v[1]=6*(b[1]-b[0])
		for (i in 2 until n-1)
		{
			u[i]=2*(h[i]+h[i-1])-h[i-1]*h[i-1]/u[i-1]
			v[i]=6*(b[i]-b[i-1])-h[i-1]*v[i-1]/u[i-1]
		}
		z[n]=0.0
		for (i in n-2 downTo 1) z[i]=(v[i]-h[i]*z[i+1])/u[i]
		z[0]=0.0
	}
	
	fun interpolate(t:Double):Double
	{
		var i=0
		var j=n-1
		while (j-i>1)
		{
			val k=(i+j)/2
			if (x[k]>t) j=k else i=k
		}
		val dx=t-x[i]
		return y[i]+dx*(z[i]/2+dx*(z[j]-z[i])/(6*(x[j]-x[i]))+(y[j]-y[i])/(x[j]-x[i]))
	}
}